## Imports

In [1]:
import pandas as pd

## Leitura da base

### Primeira leitura para checagem de dados

In [2]:
df_users = pd.read_parquet('C:/Users/candi/OneDrive/Documentos/GitHub/case_enjoei/datasets/bronze/users.parquet')

In [3]:
df_users

address  id               email  \
0  {'city': 'kilcoole', 'geolocation': {'lat': '-...   1      john@gmail.com   
1  {'city': 'kilcoole', 'geolocation': {'lat': '-...   2  morrison@gmail.com   
2  {'city': 'Cullman', 'geolocation': {'lat': '40...   3     kevin@gmail.com   
3  {'city': 'San Antonio', 'geolocation': {'lat':...   4       don@gmail.com   
4  {'city': 'san Antonio', 'geolocation': {'lat':...   5     derek@gmail.com   
5  {'city': 'el paso', 'geolocation': {'lat': '20...   6   david_r@gmail.com   
6  {'city': 'fresno', 'geolocation': {'lat': '10....   7    miriam@gmail.com   
7  {'city': 'mesa', 'geolocation': {'lat': '50.34...   8   william@gmail.com   
8  {'city': 'miami', 'geolocation': {'lat': '40.1...   9      kate@gmail.com   
9  {'city': 'fort wayne', 'geolocation': {'lat': ...  10    jimmie@gmail.com   

    username      password                                             name  \
0      johnd       m38rmF$         {'firstname': 'john', 'lastname': 'doe'}   
1   mor_2314        83r5^_   {'firstname': 'david', 'lastname': 'morrison'}   
2  kevinryan     kev02937@       {'firstname': 'kevin', 'lastname': 'ryan'}   
3     donero        ewedon        {'firstname': 'don', 'lastname': 'romer'}   
4      derek      jklg*_56     {'firstname': 'derek', 'lastname': 'powell'}   
5    david_r      3478*#54    {'firstname': 'david', 'lastname': 'russell'}   
6     snyder      f238&@*$    {'firstname': 'miriam', 'lastname': 'snyder'}   
7    hopkins  William56$hj  {'firstname': 'william', 'lastname': 'hopkins'}   
8     kate_h      kfejk@*_        {'firstname': 'kate', 'lastname': 'hale'}   
9   jimmie_k     klein*#%*     {'firstname': 'jimmie', 'lastname': 'klein'}   

            phone  __v  
0  1-570-236-7033    0  
1  1-570-236-7033    0  
2  1-567-094-1345    0  
3  1-765-789-6734    0  
4  1-956-001-1945    0  
5  1-678-345-9856    0  
6  1-123-943-0563    0  
7  1-478-001-0890    0  
8  1-678-456-1934    0  
9  1-104-001-4567    0

# Tratamentos da base

Remoção dos valores aninhados da coluna "name" e os transformando em uma nova coluna

In [4]:
# extração do primeiro e último nome guardado no dicionário para concatená-los
df_users['full_name'] = df_users['name'].apply(lambda x: f"{x['firstname']} {x['lastname']}")

### Conferência do resultado

In [5]:
df_users

address  id               email  \
0  {'city': 'kilcoole', 'geolocation': {'lat': '-...   1      john@gmail.com   
1  {'city': 'kilcoole', 'geolocation': {'lat': '-...   2  morrison@gmail.com   
2  {'city': 'Cullman', 'geolocation': {'lat': '40...   3     kevin@gmail.com   
3  {'city': 'San Antonio', 'geolocation': {'lat':...   4       don@gmail.com   
4  {'city': 'san Antonio', 'geolocation': {'lat':...   5     derek@gmail.com   
5  {'city': 'el paso', 'geolocation': {'lat': '20...   6   david_r@gmail.com   
6  {'city': 'fresno', 'geolocation': {'lat': '10....   7    miriam@gmail.com   
7  {'city': 'mesa', 'geolocation': {'lat': '50.34...   8   william@gmail.com   
8  {'city': 'miami', 'geolocation': {'lat': '40.1...   9      kate@gmail.com   
9  {'city': 'fort wayne', 'geolocation': {'lat': ...  10    jimmie@gmail.com   

    username      password                                             name  \
0      johnd       m38rmF$         {'firstname': 'john', 'lastname': 'doe'}   
1   mor_2314        83r5^_   {'firstname': 'david', 'lastname': 'morrison'}   
2  kevinryan     kev02937@       {'firstname': 'kevin', 'lastname': 'ryan'}   
3     donero        ewedon        {'firstname': 'don', 'lastname': 'romer'}   
4      derek      jklg*_56     {'firstname': 'derek', 'lastname': 'powell'}   
5    david_r      3478*#54    {'firstname': 'david', 'lastname': 'russell'}   
6     snyder      f238&@*$    {'firstname': 'miriam', 'lastname': 'snyder'}   
7    hopkins  William56$hj  {'firstname': 'william', 'lastname': 'hopkins'}   
8     kate_h      kfejk@*_        {'firstname': 'kate', 'lastname': 'hale'}   
9   jimmie_k     klein*#%*     {'firstname': 'jimmie', 'lastname': 'klein'}   

            phone  __v        full_name  
0  1-570-236-7033    0         john doe  
1  1-570-236-7033    0   david morrison  
2  1-567-094-1345    0       kevin ryan  
3  1-765-789-6734    0        don romer  
4  1-956-001-1945    0     derek powell  
5  1-678-345-9856    0    david russell  
6  1-123-943-0563    0    miriam snyder  
7  1-478-001-0890    0  william hopkins  
8  1-678-456-1934    0        kate hale  
9  1-104-001-4567    0     jimmie klein

Remoção dos valores guardados em dicionários na coluna "address" e os transformando em um dataframe

In [6]:
# Remoção dos valores de "address" e transformando em um Dataframe
df_address = pd.json_normalize(df_users['address'].tolist(),sep='_')

### Conferência do resultado

In [7]:
df_address

city  number            street     zipcode geolocation_lat  \
0     kilcoole    7682          new road  12926-3874        -37.3159   
1     kilcoole    7267         Lovers Ln  12926-3874        -37.3159   
2      Cullman      86        Frances Ct  29567-1452         40.3467   
3  San Antonio    6454  Hunters Creek Dr  98234-1734         50.3467   
4  san Antonio     245          adams St  80796-1234         40.3467   
5      el paso     124       prospect st  12346-0456         20.1677   
6       fresno    1342         saddle st  96378-0245         10.3456   
7         mesa    1342     vally view ln  96378-0245         50.3456   
8        miami     345      avondale ave  96378-0245        40.12456   
9   fort wayne     526      oak lawn ave  10256-4532        30.24788   

  geolocation_long  
0          81.1496  
1          81.1496  
2         -30.1310  
3         -20.1310  
4         -40.1310  
5         -10.6789  
6          20.6419  
7          10.6419  
8          20.5419  
9       -20.545419

Remoção das colunas address e nome + junção dos dataframes

In [8]:
# Remoção da coluna address e name originais 
df_users = df_users.drop('address', axis=1)
df_users = df_users.drop('name', axis=1)
# Juntação dos DataFrames
df_final = pd.concat([df_users, df_address], axis=1)

## Alteração e padronização dos nomes das colunas "id" e "zipcode"

In [9]:
df_final.rename(columns={'id': 'user_id', 'zipcode': 'zip_code'}, inplace=True)

### Conferência do resultado

In [10]:
df_final

user_id               email   username      password           phone  __v  \
0        1      john@gmail.com      johnd       m38rmF$  1-570-236-7033    0   
1        2  morrison@gmail.com   mor_2314        83r5^_  1-570-236-7033    0   
2        3     kevin@gmail.com  kevinryan     kev02937@  1-567-094-1345    0   
3        4       don@gmail.com     donero        ewedon  1-765-789-6734    0   
4        5     derek@gmail.com      derek      jklg*_56  1-956-001-1945    0   
5        6   david_r@gmail.com    david_r      3478*#54  1-678-345-9856    0   
6        7    miriam@gmail.com     snyder      f238&@*$  1-123-943-0563    0   
7        8   william@gmail.com    hopkins  William56$hj  1-478-001-0890    0   
8        9      kate@gmail.com     kate_h      kfejk@*_  1-678-456-1934    0   
9       10    jimmie@gmail.com   jimmie_k     klein*#%*  1-104-001-4567    0   

         full_name         city  number            street    zip_code  \
0         john doe     kilcoole    7682          new road  12926-3874   
1   david morrison     kilcoole    7267         Lovers Ln  12926-3874   
2       kevin ryan      Cullman      86        Frances Ct  29567-1452   
3        don romer  San Antonio    6454  Hunters Creek Dr  98234-1734   
4     derek powell  san Antonio     245          adams St  80796-1234   
5    david russell      el paso     124       prospect st  12346-0456   
6    miriam snyder       fresno    1342         saddle st  96378-0245   
7  william hopkins         mesa    1342     vally view ln  96378-0245   
8        kate hale        miami     345      avondale ave  96378-0245   
9     jimmie klein   fort wayne     526      oak lawn ave  10256-4532   

  geolocation_lat geolocation_long  
0        -37.3159          81.1496  
1        -37.3159          81.1496  
2         40.3467         -30.1310  
3         50.3467         -20.1310  
4         40.3467         -40.1310  
5         20.1677         -10.6789  
6         10.3456          20.6419  
7         50.3456          10.6419  
8        40.12456          20.5419  
9        30.24788       -20.545419

In [11]:
# Deletando a coluna __v por não haver uso para a mesma
df_final = df_final.drop('__v', axis=1)

## Visualização dos dados nulos

In [12]:
# Soma de todos os valores nulos por coluna
df_final.isnull().sum()

user_id             0
email               0
username            0
password            0
phone               0
full_name           0
city                0
number              0
street              0
zip_code            0
geolocation_lat     0
geolocation_long    0
dtype: int64

No caso dessa base, não houve a presença de dados nulos

## Conferência de duplicatas

A verificação foi feita nas colunas necessarias, mas deixei apenas algumas listadas a baixo como exemplo.

Vale ressaltar que algumas colunas, como "name" e "city", terão valores repeditos, mas não necessariamente são duplicatas.
Outro ponto como exemplo a se avaliar são as colunas de "password" e "username", que não devem seguir um padrão pois os valores podem ser aleatórios.

Exemplo disso é a coluna "number", que se refere a o número da residência do usuario. Ela possue dois numeros iguais, "1342", mas não necessariamente são duplicatas, pois os endereços são diferentes.
O ideal sempre é investigar os casos individualmente

### Coluna "user_id": zero duplicatas

In [14]:
# Contagem de duplicatas
df_final['user_id'].duplicated().sum()

0

### Coluna "email": zero duplicatas

In [15]:
df_final['email'].duplicated().sum()

0

### Coluna "phone": uma duplicata

In [ ]:
df_final['phone'].duplicated().sum()

1

In [ ]:
# Verificação duplicatas com base em uma ou mais colunas
df_final[df_final.duplicated(['phone'])]

user_id               email  username password           phone  \
1        2  morrison@gmail.com  mor_2314   83r5^_  1-570-236-7033   

        full_name      city  number     street    zip_code geolocation_lat  \
1  david morrison  kilcoole    7267  Lovers Ln  12926-3874        -37.3159   

  geolocation_long  
1          81.1496

In [ ]:
# Verificação duplicatas com base um valor
df_final[df_final['phone'] == '1-570-236-7033']

user_id               email  username password           phone  \
0        1      john@gmail.com     johnd  m38rmF$  1-570-236-7033   
1        2  morrison@gmail.com  mor_2314   83r5^_  1-570-236-7033   

        full_name      city  number     street    zip_code geolocation_lat  \
0        john doe  kilcoole    7682   new road  12926-3874        -37.3159   
1  david morrison  kilcoole    7267  lovers ln  12926-3874        -37.3159   

  geolocation_long  
0          81.1496  
1          81.1496

Pode haver a possibilidade de duas contas possuírem o mesmo número de telefone. É o caso do número "1-570-236-7033", onde dois usuários na base de dados possuem esse mesmo número. Além disso, algumas informações são semelhantes, como cidade, código postal e geolocalização, enquanto o número da casa e a rua diferem. Essa situação abre margem para investigação sobre se esses usuários são parentes, por morarem próximos, ou se alguém usou o número indevidamente, seja com má intenção ou de forma errônea. No entanto, aplicando uma regra de negócio que permita apenas um número por usuário, esse tipo de situação poderia ser evitado ou um possível erro identificado.

## Conferêcia de padrão nos valores

A verificação foi feita nas colunas necessarias, mas deixei apenas essas listadas a baixo como exemplo.

### Coluna "email": nada de anormal

In [35]:
df_final['email'].unique()

array(['john@gmail.com', 'morrison@gmail.com', 'kevin@gmail.com',
       'don@gmail.com', 'derek@gmail.com', 'david_r@gmail.com',
       'miriam@gmail.com', 'william@gmail.com', 'kate@gmail.com',
       'jimmie@gmail.com'], dtype=object)

### Coluna "phone": nada de anormal

In [23]:
df_final['phone'].unique()

array(['1-570-236-7033', '1-567-094-1345', '1-765-789-6734',
       '1-956-001-1945', '1-678-345-9856', '1-123-943-0563',
       '1-478-001-0890', '1-678-456-1934', '1-104-001-4567'], dtype=object)

### Coluna "city": Despadronização no valor "San Antonio" e espaços entre os nomes

In [24]:
df_final['city'].unique()

array(['kilcoole', 'Cullman', 'San Antonio', 'san Antonio', 'el paso',
       'fresno', 'mesa', 'miami', 'fort wayne'], dtype=object)

### Coluna "street": Despadronização das iniciais e espaços entre os nomes

In [25]:
df_final['street'].unique()

array(['new road', 'Lovers Ln', 'Frances Ct', 'Hunters Creek Dr',
       'adams St', 'prospect st', 'saddle st', 'vally view ln',
       'avondale ave', 'oak lawn ave'], dtype=object)

### Coluna "full_name": nada de anormal

In [27]:
df_final['full_name'].unique()

array(['john doe', 'david morrison', 'kevin ryan', 'don romer',
       'derek powell', 'david russell', 'miriam snyder',
       'william hopkins', 'kate hale', 'jimmie klein'], dtype=object)

### Padronização de valores e substituição dos espaços

In [28]:
# str.lower(): converte todo texto para minúsculo
# str.replace(' ', '_'): substitui espaços por valores escolhidos

df_final['city'] = df_final['city'].str.lower()
df_final['street'] = df_final['street'].str.lower()

### Conferência dos resultados

In [29]:
df_final['street'].unique()

array(['new road', 'lovers ln', 'frances ct', 'hunters creek dr',
       'adams st', 'prospect st', 'saddle st', 'vally view ln',
       'avondale ave', 'oak lawn ave'], dtype=object)

In [30]:
df_final['city'].unique()

array(['kilcoole', 'cullman', 'san antonio', 'el paso', 'fresno', 'mesa',
       'miami', 'fort wayne'], dtype=object)

### Dataframe com as mudanças aplicadas

In [31]:
df_final

user_id               email   username      password           phone  \
0        1      john@gmail.com      johnd       m38rmF$  1-570-236-7033   
1        2  morrison@gmail.com   mor_2314        83r5^_  1-570-236-7033   
2        3     kevin@gmail.com  kevinryan     kev02937@  1-567-094-1345   
3        4       don@gmail.com     donero        ewedon  1-765-789-6734   
4        5     derek@gmail.com      derek      jklg*_56  1-956-001-1945   
5        6   david_r@gmail.com    david_r      3478*#54  1-678-345-9856   
6        7    miriam@gmail.com     snyder      f238&@*$  1-123-943-0563   
7        8   william@gmail.com    hopkins  William56$hj  1-478-001-0890   
8        9      kate@gmail.com     kate_h      kfejk@*_  1-678-456-1934   
9       10    jimmie@gmail.com   jimmie_k     klein*#%*  1-104-001-4567   

         full_name         city  number            street    zip_code  \
0         john doe     kilcoole    7682          new road  12926-3874   
1   david morrison     kilcoole    7267         lovers ln  12926-3874   
2       kevin ryan      cullman      86        frances ct  29567-1452   
3        don romer  san antonio    6454  hunters creek dr  98234-1734   
4     derek powell  san antonio     245          adams st  80796-1234   
5    david russell      el paso     124       prospect st  12346-0456   
6    miriam snyder       fresno    1342         saddle st  96378-0245   
7  william hopkins         mesa    1342     vally view ln  96378-0245   
8        kate hale        miami     345      avondale ave  96378-0245   
9     jimmie klein   fort wayne     526      oak lawn ave  10256-4532   

  geolocation_lat geolocation_long  
0        -37.3159          81.1496  
1        -37.3159          81.1496  
2         40.3467         -30.1310  
3         50.3467         -20.1310  
4         40.3467         -40.1310  
5         20.1677         -10.6789  
6         10.3456          20.6419  
7         50.3456          10.6419  
8        40.12456          20.5419  
9        30.24788       -20.545419

## Salvando na camada prata

In [32]:
df_final.to_parquet('C:/Users/candi/OneDrive/Documentos/GitHub/case_enjoei/datasets/silver/users.parquet', index=False)